src/ProtocolZoo/ProtocolZoo.jl

```diff
@@ -0,0 +1,172 @@
+    """time-and-schedule-tracking instance from `ConcurrentSim`"""
+    sim::Simulation # TODO check that
+    """a network graph of registers"""
+    net::RegisterNet
+    """the vertex index of node A"""
+    nodeA::Int
+    """the vertex index of node B"""
+    nodeB::Int
+    """the state being generated (supports symbolic, numeric, noisy, and pure)"""
+    pairstate = StabilizerState("ZZ XX")
+    """success probability of one attempt of entanglement generation"""
+    success_prob::Float64 = 0.001
+    """duration of single entanglement attempt"""
+    attempt_time::Float64 = 0.001
+    """fixed "busy time" duration immediately before starting entanglement generation attempts"""
+    local_busy_time_pre::Float64 = 0.0
+    """fixed "busy time" duration immediately after the a successful entanglement generation attempt"""
+    local_busy_time_post::Float64 = 0.0
+    """how long to wait before retrying to lock qubits if no qubits are available (`nothing` for queuing up)"""
+    retry_lock_time::LT = 0.1
+    """how many rounds of this protocol to run (`-1` for infinite))"""
+    rounds::Int = -1
+end
+
+"""Convenience constructor for specifying `rate` of generation instead of success probability and time"""
+function EntanglerProt(sim::Simulation, net::RegisterNet, nodeA::Int, nodeB::Int; rate::Union{Nothing,Float64}=nothing, kwargs...)
+    if isnothing(rate)
+        return EntanglerProt(;sim, net, nodeA, nodeB, kwargs...)
+    else
+        return EntanglerProt(;sim, net, nodeA, nodeB, kwargs..., success_prob=0.001, attempt_time=0.001/rate)
+    end
+end
+
+#TODO """Convenience constructor for specifying `fidelity` of generation instead of success probability and time"""
+
+@resumable function (prot::EntanglerProt)(;_prot::EntanglerProt=prot)
+    prot = _prot # weird workaround for no support for `struct A a::Int end; @resumable function (fa::A) return fa.a end`; see https://github.com/JuliaDynamics/ResumableFunctions.jl/issues/77
+    rounds = prot.rounds
+    while rounds == -1 || rounds > 0
+        a = findfreeslot(prot.net[prot.nodeA])
+        b = findfreeslot(prot.net[prot.nodeB])
+        if isnothing(a) || isnothing(b)
+            isnothing(prot.retry_lock_time) && error("we do not yet support waiting on register to make qubits available") # TODO
+            @yield timeout(prot.sim, prot.retry_lock_time)
+            continue
+        end
+        @yield lock(a) & lock(b) # this yield is expected to return immediately
+        @yield timeout(prot.sim, prot.local_busy_time_pre)
+        @yield timeout(prot.sim, (rand(Geometric(prot.success_prob))+1) * prot.attempt_time)
+        initialize!((a,b), prot.pairstate; time=now(prot.sim))
+        @yield timeout(prot.sim, prot.local_busy_time_post)
+        tag!(a, :EntanglementCounterpart, prot.nodeB, b.idx)
+        tag!(b, :EntanglementCounterpart, prot.nodeA, a.idx)
+        unlock(a)
+        unlock(b)
+        rounds==-1 || (rounds -= 1)
```

Comment:
When `rounds==-1` , but the entanglement has been successfully established, how will the while loop terminate? Or when `rounds` is a finite number and entanglement was established in the first iteration, would it still keep executing for all rounds? Or is it controlled from the simulation in some way?

Reply:
The loop would not terminate, instead the protocol will continue searching for available free slots (lines 72 and 73) and if there are no free slots it will wait a bit and try again (lines 74-78).

Related: it would be nice to implement some form of `querywait` process that waits for an event to happen before running the query (in order to remove the `@yield timeout` step and instead it would just have `@yield something_changed_so_now_you_can_try_the_query_again`).

Reply:
I see, so the entangler process is always running in the background and entangling whenever a free slot is available at both the nodes.
Earlier, my understanding was that we only want a single entangled pair between each of the registers, but instead we want as many as possible and as often as possible

Reply:
if `rounds=1` it will indeed try to do it only once and then it will exit the while loop. The default of `-1` is the one that leads to infinitely trying (hopefully... if I have not done something silly)

Reply:
oh, and I guess the source of confusion is whether we are talking about "rounds of attempting entanglement" or "rounds of successfully generating entanglement". Here it is meant as the latter.